### PR TITLE
improve schema commands, testing

### PR DESCRIPTION
**Pull request by echo-bravo-yahoo**
this commit:
  - adds several tests for schema commands
  - adds mocha root hooks to clean up sinon
  - adds sinon-chai for better assertion syntax
  - swaps @cloudcmd/stub for sinon in most places
  - rewrites schema pull helpers to parallelize work

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -33,6 +33,7 @@
     "prettier": "^2.3.0",
     "rate-limiter-flexible": "^2.3.6",
     "sinon-called-with-diff": "^3.1.1",
+    "sinon-chai": "^4.0.0",
```
Comment (echo-bravo-yahoo) [inline package.json:36]: this adds some new syntactic sugar for assertions:
`expect(stub.calledWith(args)).to.be.true` => `expect(stub).to.have.been.calledWith(args)`
and
`expect(stub.called).to.be.true` => `expect(stub).to.have.been.called`

File: src/cli.mjs [hunk 2]
```diff
@@ -86,11 +86,9 @@ function buildYargs(argvInput) {
...
       const message = `${chalk.reset(yargs.help())}\n\n${chalk.red(msg || err?.message)}`
       logger.stderr(message)
-      // for some reason, this causes 2 promise rejections to be printed?
-      // debug by using `fauna reject`
-      // if (err && err.stack) {
-      //   logger.fatal(err.stack)
-      // }
+      if (err && err.stack) {
+        logger.fatal(err.stack)
+      }
```
Comment (echo-bravo-yahoo) [inline src/cli.mjs:91]: couldn't repro the issue that caused me to comment these out, so uncommenting. there's a test suite for error handling that should help us navigate these sorts of issues.

File: src/lib/schema.mjs [hunk 3]
```diff
@@ -70,6 +71,16 @@ export async function gatherRelativeFSLFilePaths(dir) {
...
 
+export async function deleteUnusedSchemaFiles(dir, filesToDelete) {
+  const fs = container.resolve("fs")
+  const promises = []
+  for (const fileName of filesToDelete) {
+    promises.push(fs.unlink(path.join(dir, fileName)))
+  }
+
+  return Promise.all(promises)
+}
```
Comment (echo-bravo-yahoo) [inline src/lib/schema.mjs:82]: parallelized re-implementation of one of the schema helpers (you'll see this pattern a few times).

File: src/yargs-commands/eval.mjs [hunk 4]
```diff
@@ -183,7 +183,7 @@ async function doEval(argv) {
       version: argv.version,
       argv
     })).client
-    : await (container.resolve("getSimpleClient")(argv))
+    : container.resolve("getSimpleClient")(argv)
```
Comment (echo-bravo-yahoo) [inline src/yargs-commands/eval.mjs:186]: awaiting the container resolution isn't necessary - i cleaned it up where i found it

File: src/yargs-commands/schema/pull.mjs [hunk 5]
```diff
@@ -53,48 +56,41 @@ async function doPull(argv) {
...
   }
   for (const overwrite of overwrites) {
-    console.log(`overwrite: ${overwrite}`);
-  }
-
-  if (argv.delete) {
-    // Delete extra .fsl files.
-      for (const deleteme of deletes) {
-        fs.unlinkSync(path.join(argv.dir, deleteme));
-      }
```
Comment (echo-bravo-yahoo) [inline src/yargs-commands/schema/pull.mjs:73]: delete used to occur _before the confirmation prompt!_

moved it to after.

File: src/yargs-commands/schema/pull.mjs [hunk 6]
```diff
@@ -53,48 +56,41 @@ async function doPull(argv) {
...
+    // the remote schema files
+    const promises = []
+    promises.push(writeSchemaFiles(contents))
+    if (argv.delete) {
+      const deleteUnusedSchemaFiles = container.resolve("deleteUnusedSchemaFiles")
+      promises.push(deleteUnusedSchemaFiles(argv.dir, deletes))
     }
+
+    // process writes and deletes together async - it'll be fastest
+    await Promise.all(promises)
```
Comment (echo-bravo-yahoo) [inline src/yargs-commands/schema/pull.mjs:92]: should be notably faster on large FSL collections now!

File: src/config/setup-container.mjs [hunk 7]
```diff
@@ -54,8 +60,12 @@ export const injectables = {
   // feature-specific lib (homemade utilities)
```
Comment (mwilde345) [inline src/config/setup-container.mjs:60]: Should these always go here? Or should the files that define them also inject them. I can imagine this list of imports and registrations gets pretty big over time. I guess I'm looking for something like spring where you can annotate the method where it's defined, and it gets registered automatically. Any pattern for that?

Comment (echo-bravo-yahoo) [reply src/config/setup-container.mjs:60]: awilix does have auto-discovery tooling; i haven't played with it any yet.

**Review by mwilde345 — APPROVED** [review]

**Pull request merged**